### PR TITLE
introduce connection params

### LINKED ISSUE #127: Add timeouts and concurrent connection limits to http client
Add reasonable default timeouts and concurrent connection limits to http client + possibility to override these values by client.

### PR DESCRIPTION
**Pull request by pirkogdc**
rewritten as agreed, closes #127

### CODE REVIEW

**Comment by martiner** [comment]
This is a bit overcomplicated and and worse - it exposes the internal structures to the client code.

**Comment by pirkogdc** [comment]
Well it was meant as preview idea for discussion.

At the second look, you are true the `GoodDataSetting.apply` is bad design - would it be better simplified to one configuration object? Or would it be still too complex for you? The reason for configuration object was to avoid exponential constructor expansion and keeping creation of GoodData simple for user.

We can also rethink the way how GoodData is created to be more flexible to future changes - may be introduce some builder?

**Comment by martiner** [comment]
I would go with a simple java bean with setters (with some smart defaults).

**Comment by pirkogdc** [comment]
Ok I will rewrite it that way

**Comment by martiner** [comment]
Are we there yet? `GoodDataSetting.apply()` is still there and it still exposes internal structures to the client code.

**Comment by pirkogdc** [comment]
I moved the `apply()` logic to `GoodData` class.

File: src/main/java/com/gooddata/GoodData.java [hunk 1]
```diff
@@ -179,6 +223,17 @@ private RestTemplate createRestTemplate(String login, String password, String ho
         return restTemplate;
     }
 
+    private void applySettings(final GoodDataSettings settings, final HttpClientBuilder clientBuilder) {
```
Comment (martiner) [inline src/main/java/com/gooddata/GoodData.java]: please rename the method to `createHttpClientBuilder` and let it return the new instance

File: src/main/java/com/gooddata/GoodData.java [hunk 2]
```diff
@@ -123,13 +165,15 @@ public GoodData(String hostname, String login, String password, int port) {
      * @param port     GoodData Platform's API port (e.g. 443)
      * @param protocol GoodData Platform's API protocol (e.g. https)
      */
-    protected GoodData(String hostname, String login, String password, int port, String protocol) {
+    protected GoodData(String hostname, String login, String password, int port, String protocol, GoodDataSettings settings) {
         notEmpty(hostname, "hostname");
         notEmpty(login, "login");
         notEmpty(password, "password");
         notEmpty(protocol, "protocol");
         final HttpClientBuilder httpClientBuilder = HttpClientBuilder.create()
```
Comment (martiner) [inline src/main/java/com/gooddata/GoodData.java]: move this to the newly created `createHttpClientBuilder` method

File: src/main/java/com/gooddata/GoodData.java [hunk 3]
```diff
@@ -179,6 +223,17 @@ private RestTemplate createRestTemplate(String login, String password, String ho
...
 
+    private void applySettings(final GoodDataSettings settings, final HttpClientBuilder clientBuilder) {
+        final PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
+        connectionManager.setDefaultMaxPerRoute(settings.getMaxConnections());
+        connectionManager.setMaxTotal(settings.getMaxConnections());
+        final RequestConfig.Builder requestConfig = RequestConfig.copy(RequestConfig.DEFAULT);
+        requestConfig.setConnectTimeout(settings.getConnectionTimeout());
+        requestConfig.setSocketTimeout(settings.getSocketTimeout());
+
+        clientBuilder.setConnectionManager(connectionManager).setDefaultRequestConfig(requestConfig.build());
```
Comment (martiner) [inline src/main/java/com/gooddata/GoodData.java]: please one `set...()` call per line

Comment (pirkogdc) [inline src/main/java/com/gooddata/GoodData.java]: done

File: src/main/java/com/gooddata/GoodData.java (hunk 2 above)
Comment (pirkogdc) [inline src/main/java/com/gooddata/GoodData.java]: done

File: src/main/java/com/gooddata/GoodData.java (hunk 1 above)
Comment (pirkogdc) [inline src/main/java/com/gooddata/GoodData.java]: renamed

**Pull request merged**